### ISSUE TITLE
Can not connect to database - Key Exchange not work

### ISSUE THREAD

**Issue by TheAkki** [keepassxreboot/keepassxc-browser#1681]
Hello,
i have a new installed firefox with deleted profiles. So i start from the scratch. I use the firefox installed from ppa on kubuntu 22.04. It is not possible for me to connect firefox with keepass.

I have checked the configuration script and the binary 'keepassxc-proxy' exist. You can see the log from strace. But i am confused. In the trace you see the line `execve("/usr/bin/keepassxc-proxy", ["/usr/bin/keepassxc-proxy", "/home/theakki/.mozilla/native-me"..., "keepassxc-browser@keepassxc.org"], 0x7f7a576b8f00 /* 69 vars */) = -1 EACCES (Keine Berechtigung)`. Is this ok?

I have tried to debug the app. But i'm not good in js. So i give up.

## Debug info
Log from Extension
`
KeePassXC-Browser: Connecting to native messaging host org.keepassxc.keepassxc_browser [client.js:317:13](moz-extension://f4867c8a-3612-4ddf-9f0a-53668b34148b/background/client.js)
[Error keepass.js:270] KeePassXC-Browser - 7: Die Nachrichtenverschlüsselung ist fehlgeschlagen. Läuft KeePassXC? 2 [global.js:178:13](moz-extension://f4867c8a-3612-4ddf-9f0a-53668b34148b/common/global.js)
Content Security Policy: Die Einstellungen der Seite haben das Laden einer Ressource auf inline blockiert ("script-src"). moz-extension:22:1
[Error event.js:127] KeePassXC-Browser - 7: Die Nachrichtenverschlüsselung ist fehlgeschlagen. Läuft KeePassXC? [global.js:178:13](moz-extension://f4867c8a-3612-4ddf-9f0a-53668b34148b/common/global.js)
[Error ] KeePassXC-Browser - No content script available for this tab. [global.js:178:13](moz-extension://f4867c8a-3612-4ddf-9f0a-53668b34148b/common/global.js)
[Error ] KeePassXC-Browser - Cannot send activated_tab message: Could not establish connection. Receiving end does not exist. 2 [global.js:178:13](moz-extension://f4867c8a-3612-4ddf-9f0a-53668b34148b/common/global.js)
[Error keepass.js:270] KeePassXC-Browser - 7: Die Nachrichtenverschlüsselung ist fehlgeschlagen. Läuft KeePassXC? [global.js:178:13](moz-extension://f4867c8a-3612-4ddf-9f0a-53668b34148b/common/global.js)
[Error keepass.js:353] KeePassXC-Browser - No content script available for this tab. [global.js:178:13](moz-extension://f4867c8a-3612-4ddf-9f0a-53668b34148b/common/global.js)
KeePassXC-Browser: Connecting to native messaging host org.keepassxc.keepassxc_browser [client.js:317:13](moz-extension://f4867c8a-3612-4ddf-9f0a-53668b34148b/background/client.js)
[Error keepass.js:753] KeePassXC-Browser - 9: Schlüsselaustausch war nicht erfolgreich. [global.js:178:13](moz-extension://f4867c8a-3612-4ddf-9f0a-53668b34148b/common/global.js)
[Error keepass.js:440] KeePassXC-Browser - No content script available for this tab. [global.js:178:13](moz-extension://f4867c8a-3612-4ddf-9f0a-53668b34148b/common/global.js)
[Error init.js:66] KeePassXC-Browser - Cannot send activated_tab message: Could not establish connection. Receiving end does not exist. 2 [global.js:178:13](moz-extension://f4867c8a-3612-4ddf-9f0a-53668b34148b/common/global.js)
`

Log from strace:
`
sudo strace -f -p $(pgrep firefox) 2>&1 | grep keepass
[pid  7868] recvmsg(127, {msg_name=NULL, msg_namelen=0, msg_iov=[{iov_base="github.com/keepassxreboot/keepas"..., iov_len=4096}], msg_iovlen=1, msg_controllen=0, msg_flags=0}, MSG_DONTWAIT) = 228
[pid  7976] sendmmsg(118, [{msg_hdr={msg_name=NULL, msg_namelen=0, msg_iov=[{iov_base="\2547\1 \0\1\0\0\0\0\0\1\tkeepassxc\3org\0\0\1\0\1\0"..., iov_len=42}], msg_iovlen=1, msg_controllen=0, msg_flags=0}, msg_len=42}, {msg_hdr={msg_name=NULL, msg_namelen=0, msg_iov=[{iov_base="\331-\1 \0\1\0\0\0\0\0\1\tkeepassxc\3org\0\0\34\0\1\0"..., iov_len=42}], msg_iovlen=1, msg_controllen=0, msg_flags=0}, msg_len=42}], 2, MSG_NOSIGNAL) = 2
[pid  7976] <... recvfrom resumed>"\2547\201\200\0\1\0\1\0\0\0\1\tkeepassxc\3org\0\0\1\0\1\300"..., 2048, 0, {sa_family=AF_INET, sin_port=htons(53), sin_addr=inet_addr("127.0.0.53")}, [28 => 16]) = 58
[pid  7976] recvfrom(118, "\331-\201\200\0\1\0\1\0\0\0\1\tkeepassxc\3org\0\0\34\0\1\300"..., 65536, 0, {sa_family=AF_INET, sin_port=htons(53), sin_addr=inet_addr("127.0.0.53")}, [28 => 16]) = 70
[pid  8223] openat(AT_FDCWD, "/home/theakki/.mozilla/native-messaging-hosts/org.keepassxc.keepassxc_browser.json", O_RDONLY <unfinished ...>
[pid  8085] newfstatat(AT_FDCWD, "/usr/bin/keepassxc-proxy",  <unfinished ...>
[pid  8246] execve("/usr/bin/keepassxc-proxy", ["/usr/bin/keepassxc-proxy", "/home/theakki/.mozilla/native-me"..., "keepassxc-browser@keepassxc.org"], 0x7f7a576b8f00 /* 69 vars */) = -1 EACCES (Keine Berechtigung)
[pid  7868] recvmsg(127, {msg_name=NULL, msg_namelen=0, msg_iov=[{iov_base="github.com/keepassxreboot/keepas"..., iov_len=4096}], msg_iovlen=1, msg_controllen=0, msg_flags=0}, MSG_DONTWAIT) = 228
[pid  7868] sendmsg(63, {msg_name=NULL, msg_namelen=0, msg_iov=[{iov_base="\24\2\0\0\377\377\377\177u\0;\0\1\0\0\0\0\0\0\0\377\377\377\377F\1\0\0(\0\0\0"..., iov_len=64}, {iov_base="usr/bin/keepassxc-proxy%20owner\277"..., iov_len=500}], msg_iovlen=2, msg_controllen=0, msg_flags=0}, MSG_DONTWAIT) = 564
[pid  8318] openat(AT_FDCWD, "/home/theakki/.mozilla/firefox/lsv8qvia.default-release/extensions/keepassxc-browser@keepassxc.org.xpi", O_RDONLY <unfinished ...>
[pid  8318] openat(AT_FDCWD, "/home/theakki/.mozilla/firefox/lsv8qvia.default-release/extensions/keepassxc-browser@keepassxc.org.xpi", O_RDONLY <unfinished ...>
[pid  8318] openat(AT_FDCWD, "/home/theakki/.mozilla/firefox/lsv8qvia.default-release/extensions/keepassxc-browser@keepassxc.org.xpi", O_RDONLY <unfinished ...>
[pid  8318] openat(AT_FDCWD, "/home/theakki/.mozilla/firefox/lsv8qvia.default-release/extensions/keepassxc-browser@keepassxc.org.xpi", O_RDONLY <unfinished ...>
[pid  8318] openat(AT_FDCWD, "/home/theakki/.mozilla/firefox/lsv8qvia.default-release/extensions/keepassxc-browser@keepassxc.org.xpi", O_RDONLY <unfinished ...>
[pid  8318] openat(AT_FDCWD, "/home/theakki/.mozilla/firefox/lsv8qvia.default-release/extensions/keepassxc-browser@keepassxc.org.xpi", O_RDONLY <unfinished ...>
[pid  8318] openat(AT_FDCWD, "/home/theakki/.mozilla/firefox/lsv8qvia.default-release/extensions/keepassxc-browser@keepassxc.org.xpi", O_RDONLY <unfinished ...>
[pid  8318] openat(AT_FDCWD, "/home/theakki/.mozilla/firefox/lsv8qvia.default-release/extensions/keepassxc-browser@keepassxc.org.xpi", O_RDONLY <unfinished ...>
[pid  8318] openat(AT_FDCWD, "/home/theakki/.mozilla/firefox/lsv8qvia.default-release/extensions/keepassxc-browser@keepassxc.org.xpi", O_RDONLY <unfinished ...>
[pid  8318] openat(AT_FDCWD, "/home/theakki/.mozilla/firefox/lsv8qvia.default-release/extensions/keepassxc-browser@keepassxc.org.xpi", O_RDONLY <unfinished ...>
[pid  8318] openat(AT_FDCWD, "/home/theakki/.mozilla/firefox/lsv8qvia.default-release/extensions/keepassxc-browser@keepassxc.org.xpi", O_RDONLY <unfinished ...>
[pid  8318] openat(AT_FDCWD, "/home/theakki/.mozilla/firefox/lsv8qvia.default-release/extensions/keepassxc-browser@keepassxc.org.xpi", O_RDONLY <unfinished ...>
[pid  8318] openat(AT_FDCWD, "/home/theakki/.mozilla/firefox/lsv8qvia.default-release/extensions/keepassxc-browser@keepassxc.org.xpi", O_RDONLY <unfinished ...>
[pid  8318] openat(AT_FDCWD, "/home/theakki/.mozilla/firefox/lsv8qvia.default-release/extensions/keepassxc-browser@keepassxc.org.xpi", O_RDONLY <unfinished ...>
[pid  8318] openat(AT_FDCWD, "/home/theakki/.mozilla/firefox/lsv8qvia.default-release/extensions/keepassxc-browser@keepassxc.org.xpi", O_RDONLY <unfinished ...>
[pid  8318] openat(AT_FDCWD, "/home/theakki/.mozilla/firefox/lsv8qvia.default-release/extensions/keepassxc-browser@keepassxc.org.xpi", O_RDONLY <unfinished ...>
[pid  8318] openat(AT_FDCWD, "/home/theakki/.mozilla/firefox/lsv8qvia.default-release/extensions/keepassxc-browser@keepassxc.org.xpi", O_RDONLY <unfinished ...>
[pid  8318] openat(AT_FDCWD, "/home/theakki/.mozilla/firefox/lsv8qvia.default-release/extensions/keepassxc-browser@keepassxc.org.xpi", O_RDONLY <unfinished ...>
[pid  7864] newfstatat(AT_FDCWD, "/usr/bin/keepassxc-proxy owner", 0x7ffe05ca5290, AT_SYMLINK_NOFOLLOW) = -1 ENOENT (Datei oder Verzeichnis nicht gefunden)
[pid  7864] newfstatat(AT_FDCWD, "/usr/bin/keepassxc-proxy owner",  <unfinished ...>
[pid  7864] newfstatat(AT_FDCWD, "/usr/bin/keepassxc-proxy owner",  <unfinished ...>
[pid  8313] openat(AT_FDCWD, "/usr/bin/keepassxc-proxy owner", O_RDONLY <unfinished ...>
[pid  8311] newfstatat(AT_FDCWD, "/usr/bin/keepassxc-proxy owner",  <unfinished ...>
[pid  8311] newfstatat(AT_FDCWD, "/usr/bin/keepassxc-proxy owner",  <unfinished ...>
[pid  7868] sendmsg(124, {msg_name=NULL, msg_namelen=0, msg_iov=[{iov_base="\334\10\0\0\377\377\377\177o\0;\0\1\0\0\0\0\0\0\0\377\377\377\377)\0\0\0(\0\0\0"..., iov_len=64}, {iov_base="keepassxc-proxy%20owner\277\0\0\0\0\4\0\0\0"..., iov_len=2236}], msg_iovlen=2, msg_controllen=0, msg_flags=0}, MSG_DONTWAIT <unfinished ...>
[pid  8319] newfstatat(AT_FDCWD, "/usr/bin/keepassxc-proxy owner", {st_mode=032205567100, st_size=140017924856272, ...}, AT_SYMLINK_NOFOLLOW) = 262
`

KeePassXC - 2.7.1
KeePassXC-Browser - 1.8.1
Operating system:/Linux Kubuntu 22.04
Browser: Firefox 102.0.1 (as deb)

**Comment by varjolintu** [comment]
Can you verify that `/usr/bin/keepassxc-proxy` is not a symbolic link?

**Comment by TheAkki** [comment]
Is no symbolic link
`
ll /usr/bin/keepassxc-proxy
`
`
-rwxr-xr-x 1 root root 35208 Apr  6 19:28 /usr/bin/keepassxc-proxy*
`

**Comment by varjolintu** [comment]
The only thing I can come up with is that KeePassXC is not setting the browser integration socket to a place it's accesible. Search for the `org.keepassxc.KeePassXC.BrowserServer` file.

**Comment by TheAkki** [comment]
sudo find / | grep org.keepassxc.KeePassXC.BrowserServer
find: ‘/run/user/1026/doc’: Keine Berechtigung
/run/user/1026/org.keepassxc.KeePassXC.BrowserServer

**Comment by varjolintu** [comment]
Everything seems to be in order then. Very strange. Have you tried some other browser?

**Comment by TheAkki** [comment]
It is possible to "try" the proxy manually?

**Comment by TheAkki** [comment]
> No, i had not tried yet. Because of 'browser are snaps' in ubuntu 22.04.

Chromium as snap is also not working

**Comment by TheAkki** [comment]
Yes, but is only stream. Which commands? Can you tell me some?

**Comment by varjolintu** [comment]
It should be just enough to start it. Entering something shouldn't crash it though.

**Comment by TheAkki** [comment]
I'm using:

dpkg --list | grep keepass
ii  keepassxc                                     2.7.1-1ppa1~jammy1                          amd64        KeePass Cross-Platform Community Edition

**Comment by FexMa** [comment]
> 

I want to add me to this problem. I have the same or similar problem.
I am also on Kubuntu 22.04.
I uninstalled firefox snap and installed the from ppa.

I get the same result in strace

```
~$ sudo strace -f -p $(pgrep firefox) 2>&1 | grep keepass
[sudo] password for : 
[pid  4919] openat(AT_FDCWD, "/home/fexma/.mozilla/native-messaging-hosts/org.keepassxc.keepassxc_browser.json", O_RDONLY <unfinished ...>
[pid  4409] newfstatat(AT_FDCWD, "/usr/bin/keepassxc-proxy",  <unfinished ...>
[pid  4987] execve("/usr/bin/keepassxc-proxy", ["/usr/bin/keepassxc-proxy", "/home/fexma/.mozilla/native-mess"..., "keepassxc-browser@keepassxc.org"], 0x7f21f156f100 /* 71 vars */) = -1 EACCES (Permission denied)
```

**Comment by Kfftfuftur** [comment]
I had the same issue after installing Ubuntu 22.04.

Manually installing Firefox and using KeePassXC 2.7.1 from the ppa fixed it for me.

**Comment by floviolleau** [comment]
Hi,

I have the same issue here.
Firefox 102.0.1 from PPA (no snap)
Keepass2 2.47 from PPA
KeePassNatMsg 2.0.14.0
Proxy 0.0.9.28122

**Comment by varjolintu** [comment]
@floviolleau What proxy is that?

**Comment by droidmonkey** [comment]
Not using keepassxc at all....

**Comment by floviolleau** [comment]
> @floviolleau What proxy is that?

https://github.com/smorks/keepassnatmsg-proxy which is equivalent to /usr/bin/keepassxc-proxy

Smorks' one is run with mono (/home/user/.mozilla/native-messaging-hosts/org.keepassxc.keepassxc_browser.json):

```
{
  "name": "org.keepassxc.keepassxc_browser",
  "description": "KeepassXC integration with Firefox with Native Messaging support",
  "path" : "/home/user/.keepassnatmsg/run-proxy.sh",
  "type": "stdio",
  "allowed_extensions": [
    "keepassxc-browser@keepassxc.org"
  ]
}
```

```
cat /home/user/.keepassnatmsg/run-proxy.sh
#!/bin/bash
mono keepassnatmsg-proxy.exe
```

instead of:
```
{
    "allowed_extensions": [
        "keepassxc-browser@keepassxc.org"
    ],
    "description": "KeePassXC integration with native messaging support",
    "name": "org.keepassxc.keepassxc_browser",
    "path": "/usr/bin/keepassxc-proxy",
    "type": "stdio"
}
```

This file is created when you click on browser integration in keepassxc or in keepass2

**Comment by varjolintu** [comment]
@floviolleau With KeePassXC you must use its own proxy. We don't officially support Keepass 2 or KeePassNatMsg-Proxy.

**Comment by floviolleau** [comment]
I know but like mentionned 

> For me both integrations didn't work (keepass2 or keepassxc)

**Comment by varjolintu** [comment]
@floviolleau https://github.com/keepassxreboot/keepassxc-browser/wiki/Troubleshooting-guide

**Comment by floviolleau** [comment]
I followed the guide and I have the same issue mentionned in https://github.com/keepassxreboot/keepassxc-browser/issues/1681#issuecomment-1186080300

**Comment by varjolintu** [comment]
@floviolleau And you installed KeePassXC from..?

**Comment by floviolleau** [comment]
Ubuntu 22.04 PPAs

**Comment by varjolintu** [comment]
I tried the new Kubuntu and I had a lof of troubles getting Firefox to be actually installed from the PPA instead. Please check `about:support` to see the binary folder.

**Comment by TheAkki** [comment]
> I tried the new Kubuntu and I had a lof of troubles getting Firefox to be actually installed from the PPA instead. Please check `about:support` to see the binary folder.

For me, the binary path of firefox is /usr/lib/firefox/firefox

**Comment by floviolleau** [comment]
same for me

**Comment by floviolleau** [comment]
Hi,

I tried 2 different ways:

with KeepassXC:
```
[pid 445831] openat(AT_FDCWD, "/home/user/.mozilla/native-messaging-hosts/org.keepassxc.keepassxc_browser.json", O_RDONLY <unfinished ...>
[pid 436696] newfstatat(AT_FDCWD, "/usr/bin/keepassxc-proxy",  <unfinished ...>
[pid 449207] execve("/usr/bin/keepassxc-proxy", ["/usr/bin/keepassxc-proxy", "/home/user/.mozilla/native-me"..., "keepassxc-browser@keepassxc.org"], 0x7fa0a9ce9d00 /* 72 vars */) = -1 EACCES (Permission denied)
```

with Keepass2 (mono):
```
[pid 445230] openat(AT_FDCWD, "/home/user/.mozilla/native-messaging-hosts/org.keepassxc.keepassxc_browser.json", O_RDONLY <unfinished ...>
[pid 445230] <... read resumed>"{\n  \"name\": \"org.keepassxc.keepa"..., 276) = 276
[pid 436696] newfstatat(AT_FDCWD, "/home/user/.keepassnatmsg/run-proxy.sh",  <unfinished ...>
[pid 436697] chdir("/home/user/.keepassnatmsg") = 0
[pid 448832] execve("/home/user/.keepassnatmsg/run-proxy.sh", ["/home/user/.keepassnatmsg/run"..., "/home/user/.mozilla/native-me"..., "keepassxc-browser@keepassxc.org"], 0x7fa0bcca0b00 /* 72 vars */) = -1 EACCES (Permission denied)
```

Same issue

Any help?

**Comment by muggenhor** [comment]
I have/had the same problem, but I think I've found the cause (and a local solution). My system:
> Ubuntu 22.04.1 LTS
> Firefox 103.0.2+build1-0ubuntu0.22.04.1~mt1 from the Mozilla PPA
> KeepassXC 2.6.6+dfsg.1-1 from Ubuntu pkg repo

After noticing the permission problems above I suspected AppArmor. So I created `/etc/apparmor.d/local/usr.bin.firefox` (loaded if it exists by `/etc/apparmor.d/usr.bin.firefox`) with this line in it:
```
/usr/bin/keepassxc-proxy Uxr,
```

Then reloaded the AppArmor profiles:
```console
sudo systemctl reload apparmor
```

When restarting firefox after that I was again able to connect to my KeepassXC db.

**Comment by muggenhor** [comment]
Just did some more digging. It looks like the PPA package disables the AppArmor profile for fresh installs. But when "upgrading" (switching from Ubuntu's own repo to the PPA) it does not disable it's own profile. It's rather convoluted so I'll copy-paste the relevant snippet from the `preinst` script:
```sh
if [ "$1" = "install" ] || [ "$1" = "upgrade" ] ; then
    # Unconditionally disable AppArmor profile for Ubuntu 9.04 and under, since
    # it requires abstractions found only in 9.10 and higher.
    major=`lsb_release -r | awk '{print $2}' | cut -d '.' -f 1`
    version=`lsb_release -r | awk '{print $2}'`
    if [ "$major" -lt 10 ] && [ "$version" != "9.10" ]; then
        disable_profile
    else
        if [ "$1" = "install" ]; then
            # Disable AppArmor profile on install, unless the last profile they
            # modified is enabled.
            base=`echo $APP_PROFILE | cut -d '-' -f 1`
            last_modified=`ls -rt $APP_DIR/$base* 2>/dev/null | grep -v '\.dpkg' | tail -n1`
            if [ -s "$last_modified" ]; then
                if [ -e "$APP_DIR/disable/`basename $last_modified`" ]; then
                    disable_profile
                fi
            else
	        # Fresh install and no other firefox profiles exist, so disable.
                disable_profile
            fi
        elif [ "$1" = "upgrade" ]; then
            # Disable AppArmor on upgrade from earlier than when we first shipped
            # the profile if the user does not already have a profile defined.
            if dpkg --compare-versions "$2" lt "3.7~a1~hg20091203" ; then
                disable_profile
            fi
        fi
    fi
```

**Comment by TheAkki** [comment]
In the meantime, without searching for a solution: It works now.
There was only normal updates and now i have firefox 103.0.1.

Can this check someone too?

**Comment by schoettler** [comment]
I had:

Firefox (snap)
KeePassXC (snap).

SSH-agent won't work from KeePassXC (snap), I decided to install KeePassXC (AppImage).

This completely bugged the browser extension. It cannot connect on fresh installs, it auto-closes and re-opens my KeePassXC (AppImage). Somehow it remembers it was once connected to KeePassXC (snap)

I tried deleting everything; local profile, local caches, the extension, checked the flag to remove data on uninstall is `false`. Removed connections to all browsers from KeePassXC (AppImage).

I don't want to restart my OS and my Firefox Account. What else can I do?
